### PR TITLE
refactor: split generic workflow from OS specific workflow

### LINKED ISSUE #560: Factor our run_on_each
I'd like to make the following change. The current structure of the build function looks like this:

```python
def build(options: BuildOptions) -> None:
    temp_dir = Path(tempfile.mkdtemp(prefix='cibuildwheel'))
    built_wheel_dir = temp_dir / 'built_wheel'
    repaired_wheel_dir = temp_dir / 'repaired_wheel'

    try:
        if options.before_all:
            log.step('Running before_all...')
            env = options.environment.as_dictionary(prev_environment=os.environ)
            before_all_prepared = prepare_command(options.before_all, project='.', package=options.package_dir)
            call([before_all_prepared], shell=True, env=env)

        python_configurations = get_python_configurations(options.build_selector)

        for config in python_configurations:
            log.build_start(config.identifier)
            ...
```

I'd like to refactor it like this:

```python
def build(options: BuildOptions) -> None:
    try:
        if options.before_all:
            log.step('Running before_all...')
            env = options.environment.as_dictionary(prev_environment=os.environ)
            before_all_prepared = prepare_command(options.before_all, project='.', package=options.package_dir)
            call([before_all_prepared], shell=True, env=env)

        python_configurations = get_python_configurations(options.build_selector)

        for config in python_configurations:
            with TemporaryDirectory(prefix='cibuildwheel') as tmp_dir:
                build_on_each(config, env, Path(tmp_dir))

    except subprocess.CalledProcessError as error:
        log.error(f'Command {error.cmd} failed with code {error.returncode}. {error.stdout}')
        sys.exit(1)


def build_on_each(config, env: Dict[str, str], tmp_dir: Path) -> None:
    built_wheel_dir = temp_dir / 'built_wheel'
    repaired_wheel_dir = temp_dir / 'repaired_wheel'

    log.build_start(config.identifier)
    ...
```

This provides a new temporary directory that is always cleaned up for each build configuration, is much better for unit testing and mocking, and removes the ridiculous levels of indentation we have for 90% of the code in build and that is keeping us from applying good programming practices, like context managers (and auto formatting? ;) ). I also think it's much easier to read the try/except, and "build_on_each" is a clearly defined concept, and still reads nicely linearly. See #558.

### PR DESCRIPTION
**Pull request by mayeut**
This PR started as an attempt to close #560 in the backlog.
As mentioned in the issue, there's probably little benefit in doing just this so I went further than that just to get some ideas where we could go. This is just a draft to see if any of those ideas are worth looking into, I don't expect this PR to be merged ever.

The end game idea was to split generic stuff from OS specific stuff which could help for #78, #676, #966, #971
To help reviewing/checking if anything is of interest:
- Virtual Environments are now handled as context managers for either build or tests
- There's now a PlatformBackend abstract class which should be implemented for all that's related to call/shell/file operations. This allows to have Windows/macOS to implement a "native/direct" one, linux the DockerContainer one.
- There's now a BuilderBackend class which handles the default build workflow thus avoiding code duplication for generic stuff. It can be customized to handle OS specific workarounds, specific implementations.

As a whole:
Pros:
- Better workflow consistency between all OS (and we could enforce the same constraints on linux as on windows/macOS easily for even more consistency) with less code duplication.
- Allows easier extendability through abstract class (one might even suggest a plug-in system as an end goal)

Cons:
- Higher complexity
- Harder to see the order of specific patch/workarounds/specifics at OS level

### CODE REVIEW

File: cibuildwheel/backend.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+from .typing import PathOrStr, assert_never
+from .util import (
+    NonPlatformWheelError,
+    get_build_verbosity_extra_flags,
+    prepare_command,
+)
+from .virtualenv import VirtualEnv, VirtualEnvBase
+
+
+class BuilderBackend:
```
Comment (henryiii) [inline cibuildwheel/backend.py:18]: I think this would be a lot simpler as a dataclass.

File: cibuildwheel/backend.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            self._before_build()
+            with self.platform.tmp_dir(self.identifier + "-built-wheel") as built_wheel_dir:
+                built_wheel = self._build_wheel(built_wheel_dir)
+                wheels = self._repair_wheel(built_wheel, repaired_wheel_dir)
+            # save build environment
+            self._constraints_dict = self.venv.constraints_dict
+        return wheels
+
+
+def test_one(
```
Comment (henryiii) [inline cibuildwheel/backend.py:244]: Starting with `test_` makes this look like a pytest test.

File: cibuildwheel/platform_backend.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+from contextlib import contextmanager
+from pathlib import Path, PurePath
+from typing import Dict, Iterator, List, Optional
+
+from cibuildwheel.bashlex_eval import local_environment_executor
+from cibuildwheel.typing import Final, Literal, PathOrStr, PlatformName
+from cibuildwheel.util import call, ensure_virtualenv, shell
+
+
+class PlatformBackend:
```
Comment (henryiii) [inline cibuildwheel/platform_backend.py:14]: Shouldn't this be an ABC? (either through the ABCMeta metaclass or inheriting from abc.ABC)?

**Review by henryiii — COMMENTED** [review]
I did a quick first pass. In general, I like this direction as long as it's easy to unit test (mocking and such should be easy?). I'll need to do a second pass looking not at the diff, but at the full code, to see if it is still reasonable to track what is happening as a reader. Also to see why there's a backend and a platform backend.

I'm not sure if @joerick will like it, given https://cibuildwheel.readthedocs.io/en/stable/contributing/#design-goals, but since it's this far, I think a "final version" should be compared with the current version (probably not through a diff) to see if it looks better, to see if it looks easy to make large specific changes without breaking the other platforms, etc. (PS: to be clear, I'm not claiming to like it yet either, just positive about the direction but will need further review later.)

**Comment by henryiii** [comment]
PS: Also, regardless of the final comparison, there are some good changes here we should pull out, like more `Final` usage.

**Comment by henryiii** [comment]
By the way, I've thought about this too, and what I likely would have attempted would have been to implement smaller methods and a Protocol that would verify that a backend had the required components. Protocols are generally nicer than ABC's, though it's easier to share code with an ABC (which is not always great for reading the code, which is part of why Protocols are nice...).

**Comment by joerick** [comment]
just a quick note - i'm very curious to review this but I won't have time for a few days!

**Comment by joerick** [comment]
Thanks for this @mayeut, it's interesting! Reflections on a quick 30-min skim through...

- Reading the code itself (outside of the diff), I get a bit lost. For example, it's not very clear to me where the _actual building_ takes place. The existing codebase, you can go to the `build` function of the platform you're interested in and read it from top-to-bottom to understand what happens. I like that - especially in the context of building things, a 'build script' style of code is more suitable I think. In this PR, we lose this straight-forwardness. Of course, the existing codebase pays for that in some code duplication, but I think readability is important enough to warrant this.
- I'm surprised how 'big' BuilderBackend is. It's got a lot of functionality in there, so perhaps suffers from the age-old problems of inheritance - where execution flow gets pretty confusing. Perhaps there would be a different version of this which tried to make the platform-specific bits as a Protocol, with as small a surface-area as possible. This might be nicer. But I'm not sure how small that surface area could get...
- Maybe it's possible to retain the build-script style, but with state-less (or minimally stated) utilities for the platform-specific bits via a Protocol? I've thought about this a little... possibly yes, but I guess we'd have to have loads of `if platform == 'windows':` and `if platform == 'macos':` for the various workarounds. I'm not sure if this would be more or less readable... potentially it could be more of a mess than the existing code. 
- I know that Lines of Code is not a great benchmark for a number of reasons, but I am surprised that this increases the count, given the duplication that was eliminated. I suppose we pay for it in genericness.

**Comment by mayeut** [comment]
@joerick, @henryiii, thanks for taking the time to review this.

I'll try to address the comments but not sure when I'll be able to address all of them.
In the meantime, if there are bits that you think should be extracted from this on their own, please let me know (as @henryiii did for `Final` usage).

**Comment by mayeut** [comment]
I rebased on main. Only the last 3 commits are addressing some of the concerns raised during review.

File: cibuildwheel/platform_backend.py (hunk 3 above)
Comment (mayeut) [reply cibuildwheel/platform_backend.py]: done.

File: cibuildwheel/backend.py (hunk 2 above)
Comment (mayeut) [reply cibuildwheel/backend.py]: changed to `do_test_wheel`

File: cibuildwheel/backend.py (hunk 1 above)
Comment (mayeut) [reply cibuildwheel/backend.py:18]: I will need some time to think about what it would look like going this way.

**Comment by mayeut** [comment]
>  The existing codebase, you can go to the build function of the platform you're interested in and read it from top-to-bottom to understand what happens. I like that - especially in the context of building things, a 'build script' style of code is more suitable I think. In this PR, we lose this straight-forwardness. Of course, the existing codebase pays for that in some code duplication, but I think readability is important enough to warrant this.

@joerick, I probably went too far in `BuilderBackend`: as I was doing `_setup_build_env`, I thought it might be good for other steps to be functions as well. I removed functions for those other steps and expanded them in `BuilderBackend::run`. It still does not answer your concern but is probably better now than before in this regard.

**Comment by joerick** [comment]
There are some good ideas here, I think it was worth attempting for sure, but ultimately I don't think that the extra layers of abstraction give us enough upside in maintainability to outweigh the downsides of indirectness.

I'm cleaning up PRs, so I'll close this one for now. Ofc feel free to reopen if you'd like to continue discussion.

**Pull request closed without merging**